### PR TITLE
FLASHIda

### PR DESCRIPTION
**Pull request by JeeH-K**
# Description

FLASHIda split from PR #5015 
To be merged after PR #5688

# Checklist:
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.

# Note:
- Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

File: cmake/Windows/Contrib/Inetc/inetc.vcxproj [hunk 1]
```diff
@@ -42,42 +42,42 @@
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release Unicode|Win32'" Label="Configuration">
     <ConfigurationType>DynamicLibrary</ConfigurationType>
-    <PlatformToolset>v110</PlatformToolset>
+    <PlatformToolset>v142</PlatformToolset>
```
Comment (timosachsenberg) [inline cmake/Windows/Contrib/Inetc/inetc.vcxproj:45]: Anyone knows why the toolset is higher here? is this loaded and automatically updated by Visual Studio?

File: src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    struct
+    {
+      bool operator()(const PeakGroup &a, const PeakGroup &b) const
+      {
+        return a.getQScore() > b.getQScore();
+      }
+    } QscoreComparator_;
+
+    /// Maps that are necessary for mass exclusion
+    std::unordered_map<int, double> mz_rt_map_;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h:135]: maybe add a comment what the int and double are

File: src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    std::unordered_map<int, double> mass_rt_map_;
+    std::unordered_map<int, double> all_mass_rt_map_;
+    std::unordered_map<int, double> mass_qscore_map_;
+
+    /**
+         @brief discard peak groups using mass exclusion
+         @param spec input spectrum
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h:142]: spec is no parameter of this method - remove?

File: src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIdaBridgeFunctions.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Kyowon Jeong $
+// $Authors: Kyowon Jeong $
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/ANALYSIS/TOPDOWN/FLASHDeconvHelperStructs.h>
+#include "OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h"
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIdaBridgeFunctions.h:38]:
```suggestion
#include <OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h>
```

File: src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp [hunk 5]
```diff
@@ -0,0 +1,1127 @@
...
+        std::cout << token_string << " is used for global targeting\n";
+      }
+      else if (token_string.hasSuffix(".out"))
+      {
+        out_files.push_back(token_string);
+        std::cout << token_string << " is used for global targeting\n";
+      }
+      else
+      {
+        double num = atof(token_string.c_str());
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp:68]: just use: token_string.toDouble()

Comment (timosachsenberg) [reply src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp:68]: hmm ok I see - there is some error handling involved

File: src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp [hunk 6]
```diff
@@ -0,0 +1,1127 @@
...
+
+    filtered_peakgroups.reserve(mass_count_.size());
+    std::set<int> current_selected_mzs; // current selected mzs
+
+    std::unordered_map<int, double> new_mz_rt_map_;
+    std::unordered_map<int, double> new_mass_rt_map_;
+    std::unordered_map<int, double> new_all_mass_rt_map_;
+    std::unordered_map<int, double> new_mass_qscore_map_;
+
+    for (auto &item: mz_rt_map_)
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp:246]: a bit easier to read with:
for (auto& [m, r] : mz_rt_map)
{
  if (rt - r > rt_window) 
...
}

File: src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp [hunk 7]
```diff
@@ -0,0 +1,1127 @@
...
+      new_mass_qscore_map_[item.first] = mass_qscore_map_[item.first];
+    }
+    new_all_mass_rt_map_.swap(all_mass_rt_map_);
+    std::unordered_map<int, double>().swap(new_all_mass_rt_map_);
+
+    new_mass_qscore_map_.swap(mass_qscore_map_);
+    std::unordered_map<int, double>().swap(new_mass_qscore_map_);
+
+
+    for (int i = 0; i < 2; i++)
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp:284]: add comment what i is

File: src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp [hunk 8]
```diff
@@ -0,0 +1,1127 @@
...
+        if (i == 0)
+        { // first, select masses or m/zs outside exclusion list
+          if (mass_rt_map_.find(nominal_mass) != mass_rt_map_.end()
+              ||
+              mz_rt_map_.find(mz) != mz_rt_map_.end()
+              )
+          {
+            continue;
+          }
+        }/* else if (mz_rt_map_.find(mz) != mz_rt_map_.end() ||
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp:326]: remove the big, commented code?

File: src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp [hunk 9]
```diff
@@ -0,0 +1,1127 @@
...
+        }
+        filtered_peakgroups.push_back(pg);
+        trigger_charges.push_back(charge);
+        // current_selected_masses.insert(nominal_mass - 1);
+        current_selected_mzs.insert(mz);
+        // current_selected_masses.insert(nominal_mass + 1);
+      }
+    }
+
+    //next_rt = rt + filtered_peakgroups.size() * .4;
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp:391]: same here... dead code

File: src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp [hunk 10]
```diff
@@ -0,0 +1,1127 @@
...
+}
+}*/
+
+    deconvoluted_spectrum_.swap(filtered_peakgroups);
+    std::vector<PeakGroup>().swap(filtered_peakgroups);
+  }
+
+  /*
+
+  void FLASHIda::filterPeakGroupsUsingMassExclusion_(const MSSpectrum &spec, const int ms_level) {
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp:480]: and here

File: src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp [hunk 11]
```diff
@@ -0,0 +1,1127 @@
...
+                                       const int ms_level, const char *name)
+  {
+    auto spec = MSSpectrum();
+    for (int i = 0; i < length; i++)
+    {
+      if (ints[i] <= 0)
+      {
+        continue;
+      }
+      spec.push_back(Peak1D(mzs[i], ints[i]));
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp:1120]: maybe spec.emplace_back(mzs[i], ints[i]);

**Review by timosachsenberg — COMMENTED** [review]
looks good. only minor cleanup of dead code and ideally more documentation would be great!

File: cmake/Windows/Contrib/Inetc/inetc.vcxproj (hunk 1 above)
Comment (KyowonJeong) [reply cmake/Windows/Contrib/Inetc/inetc.vcxproj:45]: Probably by VS. Honestly I do not know what this number means.. Should I keep v110?

File: src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h (hunk 2 above)
Comment (KyowonJeong) [reply src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h:135]: Done.

File: src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h (hunk 3 above)
Comment (KyowonJeong) [reply src/openms/include/OpenMS/ANALYSIS/TOPDOWN/FLASHIda.h:142]: Sure

File: src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp (hunk 6 above)
Comment (KyowonJeong) [reply src/openms/source/ANALYSIS/TOPDOWN/FLASHIda.cpp:246]: Done

File: cmake/Windows/Contrib/Inetc/inetc.vcxproj (hunk 1 above)
Comment (timosachsenberg) [reply cmake/Windows/Contrib/Inetc/inetc.vcxproj:45]: yeah probably better to just use the original file.

**Pull request closed without merging**